### DISCUSSION TITLE
Is there a de-reference feature for the !include syntax?

### DISCUSSION THREAD

**Discussion by LazyRichard** [plantuml/plantuml#1621] [Q&A] [answered]
when using kroki or plantuml server, It seems too difficult to convert documents with !include syntax for local files.

As an example of other specification documents, in the case of openapi, several different files can be referenced by `$ref`, but tools like `@redocly/cli` can be utilized to bundle them into a single openapi document.

In this case, a bunch of scattered files are all contained in a single openapi specification document,

Does plantuml have the ability to de-reference multiple `!include` statements into a single document like this?

for instance,

- a.puml
    ```puml
    @startuml
    interface Abc {
        +String getName()
    }
    @enduml
    ```

- b.puml
    ```puml
    @startuml
    !include ./a.puml

    interface Cde {
       +String getAge
    }
    @enduml
    ```

Assuming you have the two files as above, when you dereference them, `b.puml` will look like this

```puml
@startuml
interface Abc {
    +String getName()
}

interface Cde {
   +String getAge
}
@enduml
```

This would make it more usable when utilizing kroki or plantuml servers, as you can send a request to merge files from multiple locations into one document, but I can't seem to find it.

**Comment by The-Lum** [comment, marked as answer]
Hi @LazyRichard,

For that you can use the more generic preprocessing mechanism, as:
```
java -jar plantuml.jar -preproc file.pu
```

That produces a file `file.preproc` to send to the plantuml or kroki server.

See _ref._ here:
- https://forum.plantuml.net/7944/is-it-possible-to-run-only-the-preprocessor?show=7945#a7945
- https://forum.plantuml.net/6286/view-results-of-preprocessor
- https://plantuml.com/command-line

_Enjoy,_
Regards,
Th.
